apps/astarte_appengine_api/lib/astarte_appengine_api/device/device.ex

```diff
@@ -253,6 +257,258 @@ defmodule Astarte.AppEngine.API.Device do
+
+      Queries.insert_path_into_db(
+        client,
+        device_id,
+        interface_descriptor,
+        endpoint_id,
+        path,
+        timestamp_micro,
+        div(timestamp_micro, 1000),
+        []
+      )
+
+      DataTransmitter.push_datastream(
+        realm_name,
+        device_id,
+        interface_descriptor.name,
+        path,
+        wrapped_value
+      )
+
+      {:ok,
+       %InterfaceValues{
+         data: raw_value
+       }}
+    else
+      {:error, :unexpected_value_type, expected: value_type} ->
+        Logger.warn("Unexpected value type.", tag: "unexpected_value_type")
+        {:error, :unexpected_value_type, expected: value_type}
+
+      {:error, :invalid_object_aggregation_path} ->
+        Logger.warn("Error while trying to pubblish on path for object aggregated",
+          tag: "invalid_object_aggregation_path"
+        )
+
+        {:error, :invalid_object_aggregation_path}
+
+      {:error, :mapping_not_found} ->
+        {:error, :mapping_not_found}
+
+      {:error, reason} ->
+        Logger.warn("Unhandled error while updating object interface values: #{inspect(reason)}")
+
+        {:error, reason}
+    end
+  end
+
+  def update_interface_values(
+        realm_name,
+        encoded_device_id,
+        interface,
+        no_prefix_path,
+        raw_value,
+        _params
+      ) do
+    with {:ok, client} <- Database.connect(realm_name),
+         {:ok, device_id} <- Device.decode_device_id(encoded_device_id),
+         {:ok, major_version} <- DeviceQueries.interface_version(client, device_id, interface),
+         {:ok, interface_row} <-
+           InterfaceQueries.retrieve_interface_row(client, interface, major_version),
+         {:ok, interface_descriptor} <- InterfaceDescriptor.from_db_result(interface_row),
+         {:ownership, :server} <- {:ownership, interface_descriptor.ownership},
+         path <- "/" <> no_prefix_path do
+      if interface_descriptor.aggregation == :individual do
+        update_individual_interface_values(
+          client,
+          realm_name,
+          device_id,
+          interface_descriptor,
+          path,
+          raw_value
+        )
+      else
+        update_object_interface_values(
+          client,
+          realm_name,
+          device_id,
+          interface_descriptor,
+          path,
+          raw_value
+        )
+      end
+    else
+      {:ownership, :device} ->
+        _ = Logger.warn("Invalid write (device owned).", tag: "cannot_write_to_device_owned")
+        {:error, :cannot_write_to_device_owned}
+
+      {:error, reason} ->
+        _ = Logger.warn("Error while writing to interface.", tag: "write_to_device_error")
+        {:error, reason}
+    end
+  end
+
+  defp extract_expected_types(_path, interface_descriptor, mappings) do
+    # TODO: we should probably cache this
+    Enum.flat_map(mappings, fn mapping ->
+      if mapping.interface_id == interface_descriptor.interface_id do
```

Comment:
I know this is copypasted from DUP, but here you're retrieving the mappings from the DB explcitly for a specific interface.

Hence, I think you can drop the equality check on the interface_id, and with that you can drop the `interface_descriptor` parameter and this can become a classic `Enum.map` + `Enum.into` implemented directly with a transform function passed to `Enum.into`